cg/models/downsample/downsample_data.py

```diff
@@ -0,0 +1,161 @@
+"""Model for down sampling meta data."""
+import logging
+from pathlib import Path
+
+from cg.apps.downsample.utils import case_exists_in_statusdb
+from cg.apps.housekeeper.hk import HousekeeperAPI
+from cg.constants import Priority
+from cg.store import Store
+from cg.store.models import ApplicationVersion, Family, Sample
+from cg.utils.calculations import multiply_by_million
+
+LOG = logging.getLogger(__name__)
+
+
+class DownsampleData:
+    def __init__(
+        self,
+        status_db: Store,
+        hk_api: HousekeeperAPI,
+        sample_id: str,
+        number_of_reads: float,
+        case_id: str,
+        out_dir: Path,
+    ):
+        """Initialize the downsample data and perform integrity checks.
+        Raises:
+            ValueError
+            FileExistsError
+        """
+        self.status_db: Store = status_db
+        self.housekeeper_api: HousekeeperAPI = hk_api
+        self.sample_id: str = sample_id
+        self.number_of_reads: float = number_of_reads
+        self.case_id: str = case_id
+        self.out_dir: Path = out_dir
+        self.original_sample: Sample = self.get_sample_to_downsample()
+        self.original_case: Family = self.get_case_to_downsample()
+        self.has_enough_reads: bool = self.has_enough_reads_to_downsample()
+        self.downsampled_sample: Sample = self._generate_statusdb_downsampled_sample_record()
+        self.downsampled_case: Family = self._generate_statusdb_downsampled_case()
+        LOG.info(f"Downsample Data checks completed for {self.sample_id}")
+
+    @property
+    def downsampled_sample_name(
+        self,
+    ) -> str:
+        """Return a new sample name with the number of reads to which it is down sampled in millions appended."""
+        new_name: str = f"{self.sample_id}_{self.number_of_reads}M"
+        return new_name
+
+    @property
+    def downsampled_case_name(
+        self,
+    ) -> str:
+        """Return a case name with _downsampled appended."""
+        new_name: str = f"{self.case_id}_downsampled"
+        return new_name
+
+    def get_sample_to_downsample(self) -> Sample:
+        """Check if a sample exists in StatusDB."""
+        sample: Sample = self.status_db.get_sample_by_internal_id(self.sample_id)
+        if not sample:
+            raise ValueError(f"Sample {self.sample_id} not found in StatusDB.")
+        return sample
+
+    def get_case_to_downsample(self) -> Family:
+        """Check if a case exists in StatusDB."""
+        case: Family = self.status_db.get_case_by_internal_id(self.case_id)
+        if not case:
+            raise ValueError(f"Case {self.case_id} not found in StatusDB.")
+        return case
+
+    def _generate_statusdb_downsampled_sample_record(
+        self,
+    ) -> Sample:
+        """
+        Generate a downsampled sample record for StatusDB.
+        The new sample contains the original sample internal id and meta data.
+        Raises:
```

Comment:
Again, I think this error raising should not be here but in the function where the error is originated, but this is something that we could discuss later